### DISCUSSION TITLE
3rd party RIC e2 setup decode error

### DISCUSSION THREAD

**Discussion by wchen2654** [srsran/srsRAN_Project#279] [General Help]
Hello. I'm trying to setup a third party Near-Realtime RIC with the srsRAN Project. But for my third party Near-Realtime-RIC, I'm receiving the following error:

![image](https://github.com/srsran/srsRAN_Project/assets/105763223/04674264-06b9-41ab-b94c-4829aa039294)

Do you guys have any idea on what the problem is? If it is a problem with the asn1 compiler, is there some type of source code or version for asc1 that I could implement to decode the message?

**Comment by yagoda** [comment]
Hi!


Can you please share the gnb.log?

Thanks,

Justin

**Reply by wchen2654** [reply]
[gnb.log](https://github.com/srsran/srsRAN_Project/files/13221643/gnb.log)

**Reply by wchen2654** [reply]
And just to note that the picture that I sent in the beginning is from the third party RIC logs, not the gnb logs.

**Reply by pgawlowicz** [reply]
Hi, in your gnb logs, I see that you are running with zmq RF device and without running srsUE, right?

Note that the gnb internal timers (including those of E2 agent) are triggered by the reception of samples (e.g., every 1ms there is a `tick()` function called that advances the timers). In the case of zmq RF device, you need to connect srsUE as otherwise, the samples will not be received, and the timers are not advanced. As a result, the functions that are called on timers' expiry are never called, as timers never expire.

See in your logs, that gnb is only waiting for samples:
```
...
2023-10-30T21:59:50.204659 [zmq:rx:0:0] [I] Waiting for data.
2023-10-30T21:59:50.204907 [zmq:rx:0:0] [I] Waiting for reading samples. Completed 0 of 5760 samples.
...
````

**Reply by wchen2654** [reply]
Sorry, that was my mistake for not running the UE when giving the logs. When I ran it, my third party RIC still had an error saying that it's unable to decode the message.

Here is the new gnb.log as attached below:
[gnb.log](https://github.com/srsran/srsRAN_Project/files/13253888/gnb.log)

**Reply by pgawlowicz** [reply]
Could you provide more details, so we can reproduce the error? what is the 3rd party RIC?

**Reply by wchen2654** [reply]
So the third party RIC I'm using is from the ORAN software community and the 5g core network is Open5gs.
When I tried to connect the srsRAN project to the Near Real-time RIC provided by the ORAN community, the logs of the e2term-alpha pod printed "Error 2 Decoding (unpack) E2AP PDU from RAN". 

I looked into it more, and realized that the problem is more related to the Asn1 compiler. This is due to the encoding schema for the srsRAN Project is different from the decoding schema for the Near Real-time RIC.

So, I think the main question that I'm trying to ask is whether if you guys or anyone else have the Asn1 code or configurations that is open source for decoding the messages that the srsRAN Project send. since the Asn1 compiler that the srsRAN Project uses is internal software.

**Reply by pgawlowicz** [reply]
Hi, you can decode E2 messages with Wireshark. Note that you have to use version v4.1.0 (probably you will need to build Wireshark from source). 

see here: https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html#e2ap-packet-analyzer

**Comment by yagoda** [comment]
Hey,


Can you share a PCAP of the e2 interface?

**Reply by wchen2654** [reply]
Sorry about the late response. I'm a little new to this, but how do I get the PCAP?

**Reply by yagoda** [reply]
run a tcpdump on the interface over which the RIC and GNB are connecting.

If you are using the local interface it would be "sudo tcpdump -i lo sctp -w e2.pcap"
If you are connecting the RIC on a different machine, it is whatever interface connects your gnb computer to the RIC computer, the interface would replace "lo" in the above example (check ifconfig for the name of the interface)  run this capture, then run the RIC and GNB, kill the capture and send the resultant pcap file.

**Reply by wchen2654** [reply]
Here is the e2.pcap. I had to copy and paste it to a .txt file for it to post in the github discussions
[e2.pcap.txt](https://github.com/srsran/srsRAN_Project/files/13322580/e2.pcap.txt)

**Comment by yagoda** [comment]
Hi Wen,

There do not appear to be any E2AP messages in this pcap, additionally, the
gnb.log doesn't have an e2setupResponse received message, are you sure you
are connected over SCTP correctly?

Regards,
Justin

Justin Tallon

Engineering Manager

Software Radio Systems
***@***.***
www.srs.io
Barcelona, Spain
[image: twitter] <https://twitter.com/srssystems>
[image: linkedin] <https://www.linkedin.com/in/justin-tallon-71807136/>

**Reply by wchen2654** [reply]
So I didn't have a problem connecting the Near Real-time RIC, but the RIC had trouble decoding the message from the Srsran Project, which may cause the gnb to not receive a e2SetupResponse. At the same time, are you supposed to run that command before you run the GNB? Cause if it is, I might've done it in the wrong order. 

Here is another e2ap.pcap if that is the case:
[e2.pcap.txt](https://github.com/srsran/srsRAN_Project/files/13323569/e2.pcap.txt)

**Comment by yagoda** [comment]
hey,

There are again, no E2 messages in this pcap, can you share the gnb.conf
you are using?
The RIC should be running before the gnb is run.

also, can you copy the stdout terminal contents into a text file and share
that?

Regards,
Justin



Justin Tallon

Engineering Manager

Software Radio Systems
***@***.***
www.srs.io
Barcelona, Spain
[image: twitter] <https://twitter.com/srssystems>
[image: linkedin] <https://www.linkedin.com/in/justin-tallon-71807136/>

**Reply by wchen2654** [reply]
Of course. Just a brief description it is the gnb_zmq.conf that you guys provided with a couple of modifications to connect the near real time ric.

[gnb_zmq.txt](https://github.com/srsran/srsRAN_Project/files/13323720/gnb_zmq.txt)

**Reply by pgawlowicz** [reply]
Hi, you should sniff packets on the interface connected to RIC, as now there are not e2 messages in the pcap file.

Also, I believe you should set `bind_addr` to be in the same subnet as the `addr`, so that RIC can reach it. Or does RIC routing know how to reach 192.168.0.165?

```
e2:
  enable_du_e2: true                # Enable DU E2 agent (one for each DU instance)
  e2sm_kpm_enabled: true            # Enable KPM service module
  addr: 10.96.21.129                # RIC IP address        
  port: 36422                       # RIC port
  bind_addr: 192.168.0.165          # Local ip address
```

**Comment by PauBaguer** [comment]
Hello, same problem here, basically trying to connect srsRAN Project to O-RAN SC near-RT RIC. Find attached a capture file from "e2term-alpha" pod inside the near-RT RIC. You can see that packet 5 carries the E2setupRequest at wich point the "e2term-alpha" logs show 

`{"ts":1706273414943,"crit":"ERROR","id":"E2Terminator","mdc":{"PID":"139656364553984","POD_NAME":"deployment-ricplt-e2term-alpha-84d4db76d6-q26zd","CONTAINER_NAME":"container-ricplt-e2term","SERVICE_NAME":"RIC_E2_TERM","HOST_NAME":"edge-pc-2","SYSTEM_NAME":"SEP"},"msg":"Error 2 Decoding (unpack) E2AP PDU from RAN : "}`

[E2AP-E2term.pcapng.txt](https://github.com/srsran/srsRAN_Project/files/14065138/E2AP-E2term.pcapng.txt)

**Reply by pgawlowicz** [reply]
Hi @PauBaguer, 

Which version of ORAN SC RIC do you use? We are testing with `i-release` and we can connect our E2 agent to the RIC. 

Indeed, the second part of the `E2setupRequest` (i.e., `Item 3: id-E2nodeComponentConfigAddition`) is not filled properly yet on our side, but neither Flexric nor ORAN SC RIC do not complain about it.

Piotr

**Reply by PauBaguer** [reply]
Hi @pgawlowicz ,
This was captured using `e-release`. Thanks to your feedback I will soon update to newer releases

Thanks
Pau

**Comment by kazim425** [comment]
Dear Team,

I am also having the same error but my implementation method is different.  I have configured E2SIM from https://github.com/o-ran-sc/sim-e2-interface as a standalone,i.e. there is no RAN node but simple E2sim trying to connect to RIC.  The output is attached in a file which shows the E2sim code runs successfuly but the when checking log it shows error.

The log shows the following below 'sudo kubectl logs -f -n ricplt -l app=ricplt-e2term-alpha'

"ts":1747600987214,"crit":"ERROR","id":"E2Terminator","mdc":{"PID":"139968919746304","POD_NAME":"deployment-ricplt-e2term-alpha-db9c45968-wbkz2","CONTAINER_NAME":"container-ricplt-e2term","SERVICE_NAME":"RIC_E2_TERM","HOST_NAME":"ubuntu","SYSTEM
_NAME":"SEP"},"msg":"Error 2 Decoding (unpack) E2AP PDU from RAN : "}

[ouput.txt](https://github.com/user-attachments/files/20275766/ouput.txt)

What I am trying to achieve is that the xapp is able to send some messages e.g. Hello world to the standalone E2SIM but at the moment the RIC is not working with it. I have tred all the releases of the E2SIM available in the repo but the error is same. 

![rk263ZI-R](https://github.com/user-attachments/assets/42915ea2-0d6c-4eaa-b60f-7427af59966b)
